src/layout_engine.rs

```diff
@@ -0,0 +1,206 @@
+	Percent(f32),
+	/// Percent of free space remaining in parent container.
+	PercentRemainder(f32),
+	/// Minimum size required to fit the children.
+	Inner,
+	/// Size relative to the width of this component.
+	Width,
+	/// Size relativ to the height of this component.
+	Height,
+}
+
+impl Dimension {
+	/// Converts an `AttributeValue` to a `Dimension`.
+	fn from_attribute_value(value: &AttributeValue) -> Self {
+		if let AttributeValue::TypeValue(values) = value {
+			assert_eq!(values.len(), 1, "Unexpected number of values");
+
+			if let TypeValueOrArgument::TypeValue(ref value) = values[0] {
+				let dimension = match *value {
+					TypeValue::AbsolutePx(pixels) => Dimension::AbsolutePx(pixels),
+					TypeValue::Percent(percent) => Dimension::Percent(percent),
+					TypeValue::PercentRemainder(percent) => Dimension::PercentRemainder(percent),
+					TypeValue::Inner => Dimension::Inner,
+					TypeValue::Width => Dimension::Width,
+					TypeValue::Height => Dimension::Height,
+					_ => panic!("unsupported type in layout attribute"),
+				};
+				dimension
+			}
+			else {
+				todo!("Arguments are not supported")
+			}
+		}
+		else {
+			todo!("variable parameters are not supported")
+		}
+	}
+}
+
+impl LayoutAbstractNode {
+	// TODO: when performing layout, we call this function multiple times.
+	// We should cache this data when a node is created.
+	fn layout_attributes(&self) -> LayoutAttributes {
+		match self {
+			Self::Tag(tag) => {
+				let width = tag.width().map(Dimension::from_attribute_value).unwrap_or(Dimension::Inner);
+				let height = tag.height().map(Dimension::from_attribute_value).unwrap_or(Dimension::Inner);
+
+				LayoutAttributes { width, height }
+			},
+			Self::Text(content) => {
+				// For now, we assume the text only takes up one row,
+				// and we have a monospaced font where each character is 20px wide.
+				let width = (content.len() * 20) as f32;
+				let height = 20.0;
+
+				let width = Dimension::AbsolutePx(width);
+				let height = Dimension::AbsolutePx(height);
+
+				LayoutAttributes { width, height }
+			},
+		}
+	}
+}
+
+type LayoutNode = rctree::Node<LayoutAbstractNode>;
+
+pub fn compute_layout(root: &LayoutNode) {
+	println!("Begin layout for {:?}", root);
+
+	match &*root.borrow() {
+		LayoutAbstractNode::Tag(tag) => compute_linear_layout(root, tag),
+		LayoutAbstractNode::Text(_) => {
+			// For text, we currently just use the hard-coded layout values
+			println!("Text layout: {:?}", root.borrow().layout_attributes());
+		},
+	}
+}
+
+fn compute_linear_layout(node: &LayoutNode, tag: &LayoutAbstractTag) {
+	// Extract the layout-specific attributes of this node
+	let mut attribs = node.borrow().layout_attributes();
+
+	// For the sake of example, we'll assume the main window has a fixed size
+	if tag.namespace == "window" && tag.name == "main" {
+		attribs.width = Dimension::AbsolutePx(1920.0);
+		attribs.height = Dimension::AbsolutePx(1080.0);
+	}
```

Comment:
This is also temporary. The `main.xml` file said the window would be called with a specific width and height, so for now I've taken some example values.